File: Lambda_CS_TL_algorithms/graph_dfs_bfs.py
class Queue():
    def __init__(self):
        self.queue = []
    def enqueue(self, value):
        self.queue.append(value)
    def dequeue(self):
        if self.size() > 0:
            return self.queue.pop(0)
        else:
            return None
    def size(self):
        return len(self.queue)

class Stack():
    def __init__(self):
        self.stack = []
    def push(self, value):
        self.stack.append(value)
    def pop(self):
        if self.size() > 0:
            return self.stack.pop()
        else:
            return None
    def size(self):
        return len(self.stack)

''' BFT PSEUDOCODE:
create a queue from Queue class
Enqueue the starting vertex into the queue class 
Create an empty set, call it visited
Make a while loop of while queue is not empty then:
    Dequeue (remove from head) the first element from queue
    Check if that first dequed element is not already in the visited set then put it in the visited set
    Loop through all the neighbor vertices of that dequed vertex (that was dequeud from front of queue) with for in loop
    Add all those neighber vertices into the back of the queue (for them to be put in the visited set and their neighbors to be put in the queue
    later on)
'''

''' DFT PSEUDOCODE:
Create a stack from Stack class
Push the starting vertex into the (back of the) stack
Initialize an empty set, call it visited
Make a while loop of while stack is not empty:
    pop the last vertex from the stack and set it to a variable called get_vertex
    check if get_vertex is not in visted set then print the get_vertex vertex and add it to the visited set
    Loop through the neighbor vertices of the last element popped off from the stack (get_vertex) with for in loop.
    Add all the neighbor vertices to the end of the stack (for the last elements to be traversed first and visited and printed)
'''

'''
Pseudocode:
Graph class:
1. Make a graph class and set self.vertices to an empty dictionary

add_vertex method:
1. Takes in vertex as a parameter that'll make the input parameter vertex and put it in self.vertices dictionary as a key
2. Set the value of vertex key to an empty set

add_directed_edge method:
1. Takes in v1 and v2 as two vertices to be added as key value pairs 
2. In directed edge, v1 will be the key and v2 will be the value inside a set for the v1 key.
3. Make an if statement of both v1 and v2 is in self.vertices dictionary then add v2 as a value of v1 key using the add method for python sets

add_edge method:
1. Takes in v1 and v2 as two vertices to be added as key value pairs.
2. In undirected edge, v1 and v2 will both be each other's keys and values, they're mutually connected, it's not just 1 way connection.
3. Make an if statement, check if v1 and v2 are in self.vertices dictionary. If so, then add v1 as a key to the self.vertices dictionary and add
v2 as it's value using and add method for python sets.
4. Next, add v2 in the self.vertices dictionary and add v1 as it's value using the add method for python sets. 
5. Else (if either v1 or v2 is not in the self.vertices dictionary) raise and error saying that vertex is not present.

bft method: 
Takes in starting_vertex as a parameter. 
1. Initialize a Queue() call it queue.
2. Enqueue the starting_vertex into the queue which will add the starting_vertex into the queue.
3. Initialize an empty set, call it visited
4. Have a while loop of while queue is not empty (size of queue is more than 0)
5. Inside the while loop, dequeue or remove the first element from queue and set it equal to a variable called get_vertex
6. Make an if statement of if get_vertex is not in visited set then print the get_vertex and add it to the visited set.
7. Now traverse throug the rest of the neighber vertices in the get_vertex with a for in loop and add those neighbor vertices at the back of the 
queue. While loop will continue as long as queue has elements in it. First element will keep getting taken off the queue and get added to the 
visited set until all elements are printed and added to the visited set.

dft method:
Takes in starting_vertex as a parameter.
1. Initialize an Stack() call it stack.
2. Push the starting_vertx into the back of the stack
3. Initialize an empty set, call it visited
4. Make a while loop of while stack size is more than 0 (stack is not empty)
5. Inside while loop, pop off the last element from stack and set it to a variable called get_vertex
6. Make an if statement, check if get_vertex is not in visited set then print get_vertex and add it to the visited set.
7. Inside previous if statement, loop through each neighbor vertex of the get_vertex vertex from the self.vertices dictionary which contains the
get_vertex as key with it's neighber vertices as values in a set. This is done with a for in loop and inside the for in loop, add 
each of those neighbor vertices into the back of the stack. This will pop off each neighbor vertex then add it to the visited set if not visited
already and print it and then add it's neighbor vertices into the back of the stack and explore that 3rd neighbor to add to visited and traverse
etc.
'''


class Graph:
    def __init__(self):
        self.vertices = {}

    def add_vertex(self, vertex):
        self.vertices[vertex] = set()

    def add_directed_edge(self, v1, v2):
        if v1 in self.vertices and v2 in self.vertices:
            self.vertices[v1].add(v2)
        else:
            raise IndexError('The vertex does not exist')

    def add_edge(self, v1, v2):
        if v1 in self.vertices and v2 in self.vertices:
            self.vertices[v1].add(v2)
            self.vertices[v2].add(v1)
        else:
            raise IndexError('The vertex does not exist')

    def bft(self, starting_vertex):
        que = Queue()
        que.enqueue(starting_vertex)
        visited = set()

        while que.size() > 0:
            get_vertex = que.dequeue()

            if get_vertex not in visited:
                print(get_vertex)
                visited.add(get_vertex)
                for neighbor_vertex in self.vertices[get_vertex]:
                    que.enqueue(neighbor_vertex)

    def dft(self, starting_vertex):
        stack = Stack()
        stack.push(starting_vertex)
        visited = set()

        while stack.size() > 0:
            get_vertex = stack.pop()
            if get_vertex not in visited:
                print(get_vertex)
                visited.add(get_vertex)
                for neighbor_vertex in self.vertices[get_vertex]:
                    stack.push(neighbor_vertex)


if __name__ == '__main__':
    graph = Graph()  # Instantiate your graph
    # https://github.com/LambdaSchool/Graphs/blob/master/objectives/breadth-first-search/img/bfs-visit-order.png
    graph.add_vertex(1)
    graph.add_vertex(2)
    graph.add_vertex(3)
    graph.add_vertex(4)
    graph.add_vertex(5)
    graph.add_vertex(6)
    graph.add_vertex(7)
    graph.add_directed_edge(5, 3)
    graph.add_directed_edge(6, 3)
    graph.add_directed_edge(7, 1)
    graph.add_directed_edge(4, 7)
    graph.add_directed_edge(1, 2)
    graph.add_directed_edge(7, 6)
    graph.add_directed_edge(2, 4)
    graph.add_directed_edge(3, 5)
    graph.add_directed_edge(2, 3)
    graph.add_directed_edge(4, 6)
    
    '''
    Should print:
        {1: {2}, 2: {3, 4}, 3: {5}, 4: {6, 7}, 5: {3}, 6: {3}, 7: {1, 6}}
    '''
    print(graph.vertices)
    
    '''
    Valid DFT paths:
        1, 2, 3, 5, 4, 6, 7
        1, 2, 3, 5, 4, 7, 6
        1, 2, 4, 7, 6, 3, 5
        1, 2, 4, 6, 3, 5, 7
    '''
    graph.dft(1)
    
    '''
    Valid BFT paths:
        1, 2, 3, 4, 5, 6, 7
        1, 2, 3, 4, 5, 7, 6
        1, 2, 3, 4, 6, 7, 5
        1, 2, 3, 4, 6, 5, 7
        1, 2, 3, 4, 7, 6, 5
        1, 2, 3, 4, 7, 5, 6
        1, 2, 4, 3, 5, 6, 7
        1, 2, 4, 3, 5, 7, 6
        1, 2, 4, 3, 6, 7, 5
        1, 2, 4, 3, 6, 5, 7
        1, 2, 4, 3, 7, 6, 5
        1, 2, 4, 3, 7, 5, 6
    '''
    # print(graph.bft(1))
    
    '''
    Valid DFT recursive paths:
        1, 2, 3, 5, 4, 6, 7
        1, 2, 3, 5, 4, 7, 6
        1, 2, 4, 7, 6, 3, 5
        1, 2, 4, 6, 3, 5, 7
    '''
    # graph.dft_recursive(1)
    
    '''
    Valid BFS path:
        [1, 2, 4, 6]
    '''
    # print(graph.bfs(1, 6))
    
    '''
    Valid DFS paths:
        [1, 2, 4, 6]
        [1, 2, 4, 7, 6]
    '''
    # print(graph.dfs(1, 6))


# path = []
# starting_vert = 2
# path.append(starting_vert)
# path.append(3)
# path = path + [starting_vert]
# path = path + [3]
# print(path)
# print([path])

File: Lambda_CS_TL_algorithms/graph_dfs_bfs.js
/* BFS Path pseudocode:
1. Declare a Queue class instantiation, call it queue
2. Enqueue the starting_vertex as an array into the queue
3. Declare an empty visited set
4. While queue is not empty (size of queue is more than 0)
5. Inside while loop, dequeue the first array from the queue (initially it's just starting vertex) and call it path
6. still inside while loop, get the last element/node from the path and call it get_vertex
7. still inside while loop, if get_vertex not in visited set then, if get_vertex is equal to target_value then return path, otherwise
add node element to the visited set 
8. Loop through to get the neighbors of get_vertex from the vertices dictionary
9. Inside the for in loop, make a new copy of the path array
10. Still inside for in loop, add the neighbor vertices of get_vertex to the copied path array, 1 new path array for each new neighbor
vertex of get_vertex
11. Add the new path array(s) to the back of the queue
 */

/* DFS Path pseudocode:
1. Declare a Stack class instantiatin, call it stack
2. Declare an empty set, call it visited
3. Push the starting_vertex as an array into the back of stack
4. While stack is not empty (stack size is more than 0)
5. Inside while loop, pop the last array from back of stack, name it path & get the last value from path, name it get_vertex
6. Still inside while loop, check if get_vertex is not in visited set then put it in visited set in the if statement.
7. Still inside the previous if statement of if get_vertex is not in visited then check if get_vertex is equal to target_value then 
return path. Otherwise...
8. Still inside the previous if statement of if get_vertex is not in visited, get the neighbors set of get_vertex vertex from the 
self.vertices dictionary and call it neighbor_vertices
9. Make a helper function called pathAdder where it takes in value as an argument, and inside function, we make a new copy of the path
with path.slice() method, call it new_path, push the input values into the back of the new_path then push the new_path to the back of
the stack
10. Call the pathAdder function on each vertex of the neighbor_vertices set using the forEach sets method.
 */

/* Breadth-first traversal pseudocode:
1. Declare an instantiation of the Queue class, call it queue
2. Declare an empty set, call it visited
3. Push the input starting_vertex into the back of Queue using the enqueue method
4. While queue is not empty (queue size is more than 0)
5. In the while loop, take out the vertex from the front of the queue using dequeue method and call the vertex get_vertex
6. Still inside while loop, if get_vertex vertex is not in visited set, add it to the visited set.
7. Still inside if statement of if get_vertex is not in visited set, get the set of neighbors of the get_vertex and call it neighbor_vertices
8. Make a helper function called neighborTraversal which takes in value as an input parameter. Inside neighborTraveral function enqueue the
input value onto the back of the queue so that we can then get this neighbor's neighbors and print it out
9. Outside, the neighborTraversal function, call the forEach method of the neigbor_vertices set (neighbors set of get_vertex) and pass in the
neighborTraversal helper function as a callback to that forEach method.   
 */

/* Depth first traversal recursive
1. Takes in starting_vertex and visited set to null as input parameters
2. Inside function, check if visited is None then set visited equal to an empty set
3. Add the input starting_vertex to visited set
4. Print out the starting_vertex with console log
5. Get the neighbors of starting_vertex from the self.vertices dictionary call it neighbors
6. Make a helper function called neighborAdder which will be the callback, it takes in value as an argument. Inside the neighborAdder function
check if value is not in visited then recursively call the dft_recursive outer function, passing in value as the 1st parameter (in place of 
  starting_vertex and visited set)
6. Call the neighborAdder helper function with the forEach method on the neighbor set from the starting_vertex neighbors. 
 */
class Queue {
  constructor() {
    this.queue = [];
  }
  enqueue(value) {
    this.queue.push(value);
  }

  dequeue() {
    if (this.size() > 0) {
      return this.queue.shift();
    } else {
      return null;
    }
  }

  size() {
    return this.queue.length;
  }
}

class Stack {
  constructor() {
    this.stack = [];
  }
  push(value) {
    this.stack.push(value);
  }
  pop() {
    if (this.size() > 0) {
      return this.stack.pop();
    } else {
      return null;
    }
  }

  size() {
    return this.stack.length;
  }
}

class Graph {
  constructor() {
    this.vertices = {};
  }
  add_vertex(vertex_id) {
    this.vertices[vertex_id] = new Set();
  }

  add_edge(v1, v2) {
    if (v1 in this.vertices && v2 in this.vertices) {
      this.vertices[v1].add(v2);
      // this.vertices[v2].add(v1);
    } else {
      console.log("The vertex does not exist!");
    }
  }

  add_directed_edge(v1, v2) {
    if (v1 in this.vertices && v2 in this.vertices) {
      this.vertices[v1].add(v2);
    }
  }

  bfs_path(starting_vertex_id, target_value) {
    let queue = new Queue();
    queue.enqueue([starting_vertex_id]);
    let visited = new Set();

    while (queue.size() > 0) {
      let path = queue.dequeue();
      //   console.log("path: ", path);
      let get_vertex = path[path.length - 1];
      //   console.log("get_vertex: ", get_vertex);

      if (visited.has(get_vertex) == false) {
        visited.add(get_vertex);
        // console.log("visited: ", visited);

        if (get_vertex == target_value) {
          //   console.log(path);
          return path;
        }
        // console.log('self.vertices[get_vertex]: ', this.vertices[get_vertex]);
        let neighbor_vertices = this.vertices[get_vertex];
        let pathAdder = values => {
          let new_path = path.slice();
          new_path.push(values);
          queue.enqueue(new_path);
        };
        neighbor_vertices.forEach(pathAdder);
      }
    }

    return null;
  }

  dfs_path(starting_vertex_id, target_value) {
    let stack = new Stack();
    let visited = new Set();
    stack.push([starting_vertex_id]);

    while (stack.size() > 0) {
      let path = stack.pop();
      let get_vertex = path[path.length - 1];

      if (visited.has(get_vertex) == false) {
        visited.add(get_vertex);

        if (get_vertex == target_value) {
          return path;
        }

        let neighbor_vertices = this.vertices[get_vertex];

        let pathAdder = values => {
          let new_path = path.slice();
          new_path.push(values);
          stack.push(new_path);
        };

        neighbor_vertices.forEach(pathAdder);
      }
    }

    return null;
  }

  bft(starting_vertex_id) {
    let queue = new Queue();
    let visited = new Set();
    queue.enqueue(starting_vertex_id);

    while (queue.size() > 0) {
      let get_vertex = queue.dequeue();

      if (visited.has(get_vertex) == false) {
        console.log(get_vertex);
        visited.add(get_vertex);

        let neighbor_vertices = this.vertices[get_vertex];

        let neighborTraversal = value => {
          queue.enqueue(value);
        };

        neighbor_vertices.forEach(neighborTraversal);
      }
    }
    return;
  }

  dft(starting_vertex_id) {
    let stack = new Stack();
    let visited = new Set();

    stack.push(starting_vertex_id);

    while (stack.size() > 0) {
      let get_vertex = stack.pop();

      if (visited.has(get_vertex) == false) {
        visited.add(get_vertex);
        console.log(get_vertex);

        let neigbor_vertices = this.vertices[get_vertex];

        let neighborTraversal = value => {
          stack.push(value);
        };

        neigbor_vertices.forEach(neighborTraversal);
      }
    }
  };

  dft_recursive(starting_vertex, visited=null) {
    if (visited == null) {
      visited = new Set();
    };

    visited.add(starting_vertex);
    console.log(starting_vertex);

    let neighbors = this.vertices[starting_vertex];

    let neighborAdder = value => {
      if (visited.has(value) == false) {
        this.dft_recursive(value, visited)
      }
    };

    neighbors.forEach(neighborAdder);
  }
}

let graph = new Graph();
graph.add_vertex(1);
graph.add_vertex(2);
graph.add_vertex(3);
graph.add_vertex(4);
graph.add_vertex(5);
graph.add_vertex(6);
graph.add_vertex(7);
graph.add_edge(5, 3);
graph.add_edge(6, 3);
graph.add_edge(7, 1);
graph.add_edge(4, 7);
graph.add_edge(1, 2);
graph.add_edge(7, 6);
graph.add_edge(2, 4);
graph.add_edge(3, 5);
graph.add_edge(2, 3);
graph.add_edge(4, 6);

// console.log("graph.vertices: ", graph.vertices);

/*
Valid BFS path:
    [1, 2, 4, 6]
*/
// console.log(graph.bfs_path(1, 6));

/*
Valid DFS paths:
    [1, 2, 4, 6]
    [1, 2, 4, 7, 6]
*/
// console.log(graph.dfs_path(1, 6));

/*
Valid BFT paths:
    1, 2, 3, 4, 5, 6, 7
    1, 2, 3, 4, 5, 7, 6
    1, 2, 3, 4, 6, 7, 5
    1, 2, 3, 4, 6, 5, 7
    1, 2, 3, 4, 7, 6, 5
    1, 2, 3, 4, 7, 5, 6
    1, 2, 4, 3, 5, 6, 7
    1, 2, 4, 3, 5, 7, 6
    1, 2, 4, 3, 6, 7, 5
    1, 2, 4, 3, 6, 5, 7
    1, 2, 4, 3, 7, 6, 5
    1, 2, 4, 3, 7, 5, 6
 */
// console.log(graph.bft(1));

/*
Valid DFT paths:
    1, 2, 3, 5, 4, 6, 7
    1, 2, 3, 5, 4, 7, 6
    1, 2, 4, 7, 6, 3, 5
    1, 2, 4, 6, 3, 5, 7
 */
// console.log(graph.dft(1));

/*
Valid DFT recursive paths:
    1, 2, 3, 5, 4, 6, 7
    1, 2, 3, 5, 4, 7, 6
    1, 2, 4, 7, 6, 3, 5
    1, 2, 4, 6, 3, 5, 7
 */
console.log(graph.dft_recursive(1));